core/builtin.py

```diff
@@ -604,28 +604,35 @@ def Cd(argv, mem, dir_stack):
   return 0
 
 
-WITH_PREFIX = 1
-WITHOUT_PREFIX = 2
+WITH_LINE_NUMBERS = 1
+WITHOUT_LINE_NUMBERS = 2
 SINGLE_LINE = 3
 
-def _PrintDirStack(dir_stack, style):
+def _PrintDirStack(dir_stack, style, home_dir):
   """Helper for 'dirs'."""
 
-  if style == WITH_PREFIX:
+  if style == WITH_LINE_NUMBERS:
     for i, entry in enumerate(dir_stack.Iter()):
-      print('%2d  %s' % (i, entry))
+      print('%2d  %s' % (i, _FormatDir(entry, home_dir)))
 
-  elif style == WITHOUT_PREFIX:
+  elif style == WITHOUT_LINE_NUMBERS:
     for entry in dir_stack.Iter():
-      print(entry)
+      print(_FormatDir(entry, home_dir))
 
   elif style == SINGLE_LINE:
-    print(' '.join(dir_stack.Iter()))
+    print(' '.join(_FormatDir(entry, home_dir) \
+      for entry in dir_stack.Iter()))
 
   sys.stdout.flush()
 
 
-def Pushd(argv, dir_stack):
+def _FormatDir(dir_name, home_dir):
+  if home_dir and home_dir.tag == value_e.Str and dir_name.startswith(home_dir.s):
```

Comment:
I think there is still a bug here, where

```
HOME=/home/andy
```

and

one of the entries is `/home/andyc`.  In that case it doesn't make sense to print `~c`.

So you have to handle at least two cases, and yes there should be tests for these.

Reply:
Ahh, yes I missed that case. I'll write up some tests and figure out how to handle that